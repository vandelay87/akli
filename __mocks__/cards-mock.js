export const cardMock = [
  {
    id: '1',
    title: 'Lorem Ipsum',
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
  }
];

export const cardWithSubtitleMock = [
  {
    id: '1',
    title: 'Lorem Ipsum',
    subtitle: 'Duis aute irure',
    description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
  }
];

export const cardsWithImageMock = [
  {
    id: '1',
    title: 'Facebook',
    subtitle: 'email@facebook.com',
    image: {
      file: {
        url: 'https://via.placeholder.com/200x120',
      }
    },
    description: 'Connect with friends, family and other people you know. Share photos and videos, send messages and get updates.',
  },
  {
    id: '2',
    title: 'Twitter',
    subtitle: 'email@twitter.com',
    image: {
      file: {
        url: 'https://via.placeholder.com/200x120',
      }
    },
    description: 'See what is happening in the world right now. Join Twitter today. Sign up Log In.',
  },
  {
    id: '3',
    title: 'LinkedIn',
    subtitle: 'email@linkedin.com',
    image: {
      file: {
        url: 'https://via.placeholder.com/200x120',
      }
    },
    description: 'Manage your professional identity. Build and engage with your professional network. Access knowledge, insights and opportunities.',
  },
];