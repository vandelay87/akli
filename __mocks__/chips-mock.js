// eslint-disable-next-line import/no-extraneous-dependencies
import { action } from '@storybook/addon-actions'

export const chipsWithActionMock = [
  {
    id: '1',
    value: 'Small',
    action: action('small-click'),
  },
  {
    id: '2',
    value: 'Medium',
    action: action('medium-click'),
  },
  {
    id: '3',
    value: 'Large',
    action: action('large-click'),
  },
]

export const chipsWithLinkMock = [
  {
    id: '1',
    value: 'React.js',
    link: 'https://reactjs.org/',
  },
  {
    id: '2',
    value: 'Vue',
    link: 'https://vuejs.org/',
  },
  {
    id: '3',
    value: 'Angular',
    link: 'https://angular.io/',
  },
  {
    id: '4',
    value: 'Gatsby',
    link: 'https://www.gatsbyjs.org/',
  },
]
