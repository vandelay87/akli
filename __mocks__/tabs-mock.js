export const tabsMock = [
  {
    id: '1',
    label: 'tab 1',
    content: {
      id: 'c1',
      json: {
        content: [
          {
            content: [
              {
                marks: [],
                value: 'Tab 1 content',
                nodeType: 'text',
              },
            ],
            nodeType: 'heading-2',
          },
        ],
      },
    },
  },
  {
    id: '2',
    label: 'tab 2',
    content: {
      id: 'c2',
      json: {
        content: [
          {
            content: [
              {
                marks: [],
                value: 'Tab 2 content',
                nodeType: 'text',
              },
            ],
            nodeType: 'heading-2',
          },
        ],
      },
    },
  },
  {
    id: '3',
    label: 'tab 3',
    content: {
      id: 'c3',
      json: {
        content: [
          {
            content: [
              {
                marks: [],
                value: 'Tab 3 content',
                nodeType: 'text',
              },
            ],
            nodeType: 'heading-2',
          },
        ],
      },
    },
  },
]
