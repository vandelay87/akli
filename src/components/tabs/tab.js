import React from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import { CSSTransition } from 'react-transition-group'
import Ripple from '../../hooks/ripple'
import { robotoRegular } from '../../styles/fonts'
import { color } from '../../styles/colors'

const Tab = ({
  id,
  label,
  click,
  keyDown,
  selected,
  focused,
  controls,
  timeout,
}) => (
  <StyledTab
    role="tab"
    id={id}
    onClick={click}
    onKeyDown={keyDown}
    selected={selected}
    tabIndex={focused ? 0 : -1}
    aria-selected={selected}
    aria-controls={controls}
  >
    {label}
    <Ripple color={color.primary} duration={600} />
    <CSSTransition in={selected} timeout={timeout}>
      {state => <StyledTab.Indicator state={state} timeout={timeout} />}
    </CSSTransition>
  </StyledTab>
)

const getIndicatorTransition = state => {
  switch (state) {
    case 'entering':
    case 'exited':
      return `
        transform: translateY(2px);
      `

    case 'entered':
    default:
      return `
        transform: translateY(0);
      `
  }
}
const StyledTab = styled.button`
  ${robotoRegular}
  flex: 1 0 auto;
  font-size: 0.875rem;
  line-height: 2.25rem;
  font-weight: 500;
  text-transform: uppercase;
  background: transparent;
  border-width: 0;
  border-style: none;
  color: #616161;
  overflow: hidden;
  position: relative;

  &:hover {
    background: #f6fafd;
    cursor: pointer;
  }

  &:focus {
    background: #e5effa;
    outline: none;
  }

  ${({ selected }) =>
    selected &&
    css`
      color: #1976d2;
    `}
`
StyledTab.Indicator = styled.span`
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 2px;
  background: ${color.primary};
  transition: transform ${({ timeout }) => timeout}ms ease-in-out;

  ${({ state }) =>
    css`
      ${getIndicatorTransition(state)}
    `}
`

Tab.propTypes = {
  id: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  click: PropTypes.func.isRequired,
  keyDown: PropTypes.func.isRequired,
  selected: PropTypes.bool.isRequired,
  focused: PropTypes.bool.isRequired,
  controls: PropTypes.string.isRequired,
  timeout: PropTypes.number,
}
Tab.defaultProps = {
  timeout: 200,
}

export default Tab
