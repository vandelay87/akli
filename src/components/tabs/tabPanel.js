import React from 'react'
import PropTypes from 'prop-types'
import styled, { css } from 'styled-components'
import { CSSTransition } from 'react-transition-group'
import RichText from '../richText/richText'

const TabPanel = ({ id, content, selected, labelledBy, timeout }) => (
  <CSSTransition in={selected} timeout={timeout} unmountOnExit>
    {state => (
      <StyledTabPanel
        role="tabpanel"
        id={id}
        state={state}
        aria-hidden={!selected}
        aria-labelledby={labelledBy}
        timeout={timeout}
      >
        <RichText content={content} />
      </StyledTabPanel>
    )}
  </CSSTransition>
)

const getPanelTransition = state => {
  switch (state) {
    case 'entering':
      return `
        opacity: 0;
        transform: scale(0.97) translateY(5px);
      `

    case 'entered':
    case 'exiting':
      return `
        opacity: 1;
        transform: scale(1) translateY(0);
      `

    default:
      return `
        opacity: 0;
        transform: scale(0.97) translateY(5px);
      `
  }
}
const StyledTabPanel = styled.div`
  transition: opacity ${({ timeout }) => timeout}ms linear,
    transform ${({ timeout }) => timeout}ms ease-in-out;

  ${({ state }) =>
    css`
      ${getPanelTransition(state)}
    `}
`

TabPanel.propTypes = {
  id: PropTypes.string.isRequired,
  content: PropTypes.shape({
    id: PropTypes.string,
    json: PropTypes.shape({}),
  }).isRequired,
  selected: PropTypes.bool.isRequired,
  labelledBy: PropTypes.string.isRequired,
  timeout: PropTypes.number,
}
TabPanel.defaultProps = {
  timeout: 200,
}

export default TabPanel
