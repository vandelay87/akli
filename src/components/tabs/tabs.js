import React, { useState, useRef, useEffect } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { applyStyleModifiers } from 'styled-components-modifiers'
import RichText from '../richText/richText'
import { robotoRegular } from '../../styles/fonts'
import { color } from '../../styles/colors'

const Tabs = ({ label, tabList }) => {
  const [tabs, setTabs] = useState(
    tabList.map((tab, index) => ({
      ...tab,
      selected: index ? false : true,
      focused: index ? false : true,
    }))
  )
  const tabBarEl = useRef(null)
  const didMount = useRef(false);

  useEffect(() => {
    if (didMount.current) {
      didMount.current = false
      return
    }

    const tabsEl = tabBarEl.current.children
    Object.keys(tabsEl).forEach(key => {
      tabsEl[key].attributes.tabIndex.value === '0' && tabsEl[key].focus()
    })
  }, [tabs])
  
  const setSelectedTab = selectedTab => (
    setTabs(tabs.map((tab, index) => ({
      ...tab,
      selected: selectedTab === index,
      focused: selectedTab === index,
    })))
  )

  const setFocusedTab = focusedTab => (
    setTabs(tabs.map((tab, index) => ({
      ...tab,
      focused: focusedTab === index,
    })))
  )

  const handleTabKeyPress = (event, index) => {
    if (event.key === 'ArrowLeft') {
      setFocusedTab(!index ? tabs.length - 1 : index - 1)
    }
    if (event.key === 'ArrowRight') {
      setFocusedTab(index === tabs.length - 1 ? 0 : index + 1)
    }
  }

  return (
    <StyledWrapper aria-label={label}>
      <StyledTabBar role="tablist" ref={tabBarEl}>
        {tabs.map((tab, index) => (
          <StyledTab
            role="tab"
            id={tab.id}
            onClick={() => setSelectedTab(index)}
            onKeyDown={e => handleTabKeyPress(e, index)}
            tabIndex={tab.focused ? 0 : -1}
            aria-selected={tab.selected}
            aria-controls={tab.content.id}
            key={tab.id}
          >
            {tab.label}
            <StyledTab.Indicator modifiers={tab.selected && 'selected'} />
          </StyledTab>
        ))}
      </StyledTabBar>
      {tabs.map(tab => (
        <StyledTabPanel
          role="tabpanel"
          id={tab.content.id}
          modifiers={tab.selected && 'selected'}
          aria-hidden={!tab.selected}
          aria-labelledby={tab.id}
          key={tab.content.id}
        >
          <RichText article={tab.content} />
        </StyledTabPanel>
      ))}
    </StyledWrapper>
  )
}

const TABBAR_CONFIG = {
  selected: () => `
    background: ${color.primary};
  `,
}
const TABPANEL_CONFIG = {
  selected: () => `
    display: block;
  `,
}
const StyledWrapper = styled.div`
  ${robotoRegular}
  margin: 1em auto;
`
const StyledTabBar = styled.div`
  display: flex;
  overflow-x: scroll;
  overflow-y: hidden;
  scrollbar-width: none;
  -ms-overflow-style: none;

  &::-webkit-scrollbar {
    display: none;
  }
`
const StyledTab = styled.button`
  flex: 1 0 auto;
  font-size: 0.875rem;
  line-height: 2.25rem;
  font-weight: 500;
  text-transform: uppercase;
  background: transparent;
  border-width: 0;
  border-style: none;
  color: #616161;
  overflow: hidden;
  position: relative;
  transition: all 0.3s ease-out;

  &:hover {
    background: #f6fafd;
    color: #1976d2;
    cursor: pointer;
  }

  &:focus {
    background: #e5effa;
    outline: none;
  }
`
StyledTab.Indicator = styled.span`
  position: absolute;
  bottom: 0;
  left: 0;
  width: 100%;
  height: 2px;
  background: transparent;

  ${applyStyleModifiers(TABBAR_CONFIG)};
`
const StyledTabPanel = styled.div`
  display: none;

  ${applyStyleModifiers(TABPANEL_CONFIG)};
`

Tabs.propTypes = {
  label: PropTypes.string.isRequired,
  tabList: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string,
      label: PropTypes.string,
      content: PropTypes.shape({
        id: PropTypes.string,
        json: PropTypes.json,
      }),
    })
  ).isRequired,
}

export default Tabs
