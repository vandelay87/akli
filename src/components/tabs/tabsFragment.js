import { graphql } from 'gatsby'

export const query = graphql`
  fragment Tabs on ContentfulTabs {
    id
    label
    tabList {
      id
      label
      content {
        id
        json
      }
    }
  } 
`