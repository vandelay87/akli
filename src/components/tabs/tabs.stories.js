import React from 'react'
import Tabs from './tabs'
import { tabsMock } from '../../../__mocks__/tabs-mock'

export default {
  component: Tabs,
  title: 'Tabs',
}

export const tabs = () => <Tabs label="tabs" tabList={tabsMock} />
