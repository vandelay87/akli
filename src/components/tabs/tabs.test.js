import React from 'react'
import renderer from 'react-test-renderer'
import { act } from 'react-dom/test-utils'
import { mount } from 'enzyme'
import Tabs from './tabs'
import { tabs } from './tabs.stories'
import { tabsMock } from '../../../__mocks__/tabs-mock'

describe('tabs.js', () => {
  it('renders component', () => {
    const tree = renderer.create(tabs()).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('sets selected tab', () => {
    const wrapper = mount(<Tabs label="tabs" tabList={tabsMock} />)
    let tabNodes = wrapper.find('button[role="tab"]')
    let tabpanelNodes = wrapper.find('div[role="tabpanel"]')

    expect(tabNodes.at(0).prop('aria-selected')).toBe(true)
    expect(tabNodes.at(1).prop('aria-selected')).toBe(false)
    expect(tabpanelNodes.at(0).prop('aria-hidden')).toBe(false)
    expect(tabpanelNodes.at(1).prop('aria-hidden')).toBe(true)

    act(() => {
      tabNodes.at(1).prop('onClick')()
    })
    wrapper.setProps()
    tabNodes = wrapper.find('button[role="tab"]')
    tabpanelNodes = wrapper.find('div[role="tabpanel"]')

    expect(tabNodes.at(1).prop('aria-selected')).toBe(true)
    expect(tabNodes.at(0).prop('aria-selected')).toBe(false)
    expect(tabpanelNodes.at(1).prop('aria-hidden')).toBe(false)
    expect(tabpanelNodes.at(0).prop('aria-hidden')).toBe(true)
  })

  it('sets focused tab', () => {
    const wrapper = mount(<Tabs label="tabs" tabList={tabsMock} />)
    let tabNodes = wrapper.find('button[role="tab"]')

    function handleOnKeyDown(keypress, index) {
      act(() => {
        tabNodes.at(index).prop('onKeyDown')({ key: keypress })
      })
      wrapper.setProps()
      tabNodes = wrapper.find('button[role="tab"]')
    }

    expect(tabNodes.at(0).prop('tabIndex')).toBe(0)
    expect(tabNodes.at(1).prop('tabIndex')).toBe(-1)

    handleOnKeyDown('ArrowRight', 0)
    expect(tabNodes.at(0).prop('tabIndex')).toBe(-1)
    expect(tabNodes.at(1).prop('tabIndex')).toBe(0)

    handleOnKeyDown('ArrowLeft', 1)
    expect(tabNodes.at(0).prop('tabIndex')).toBe(0)
    expect(tabNodes.at(1).prop('tabIndex')).toBe(-1)

    handleOnKeyDown('ArrowLeft', 0)
    expect(tabNodes.at(0).prop('tabIndex')).toBe(-1)
    expect(tabNodes.at(tabsMock.length - 1).prop('tabIndex')).toBe(0)

    handleOnKeyDown('ArrowRight', tabsMock.length - 1)
    expect(tabNodes.at(tabsMock.length - 1).prop('tabIndex')).toBe(-1)
    expect(tabNodes.at(0).prop('tabIndex')).toBe(0)
  })
})
