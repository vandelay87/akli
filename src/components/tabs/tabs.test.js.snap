// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tabs.js renders component 1`] = `
<article
  aria-label="tabs"
  className="sc-jzJRlG eFrhWN"
>
  <div
    className="sc-cSHVUG gwqELa"
    role="tablist"
  >
    <button
      aria-controls="c1"
      aria-selected={true}
      className="sc-bwzfXH bIKAQi"
      id="1"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      selected={true}
      tabIndex={0}
    >
      tab 1
      <div
        aria-hidden="true"
        className="sc-bdVaJa JrgfE"
        color="#1976d2"
        onMouseDown={[Function]}
      />
      <span
        className="sc-htpNat iCtHmj"
      />
    </button>
    <button
      aria-controls="c2"
      aria-selected={false}
      className="sc-bwzfXH cUvomQ"
      id="2"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      selected={false}
      tabIndex={-1}
    >
      tab 2
      <div
        aria-hidden="true"
        className="sc-bdVaJa JrgfE"
        color="#1976d2"
        onMouseDown={[Function]}
      />
      <span
        className="sc-htpNat dsHXqs"
      />
    </button>
    <button
      aria-controls="c3"
      aria-selected={false}
      className="sc-bwzfXH cUvomQ"
      id="3"
      onClick={[Function]}
      onKeyDown={[Function]}
      role="tab"
      selected={false}
      tabIndex={-1}
    >
      tab 3
      <div
        aria-hidden="true"
        className="sc-bdVaJa JrgfE"
        color="#1976d2"
        onMouseDown={[Function]}
      />
      <span
        className="sc-htpNat dsHXqs"
      />
    </button>
  </div>
  <div
    aria-hidden={false}
    aria-labelledby="1"
    className="sc-fjdhpX twTGk"
    id="c1"
    role="tabpanel"
  >
    <div
      className="sc-jTzLTM cFAJuX"
    >
      <h2
        className="sc-bxivhb bdCLac"
        size={2}
      >
        Tab 1 content
      </h2>
    </div>
  </div>
</article>
`;
